Contest_API.md

```diff
@@ -916,11 +916,11 @@ Properties of language objects:
 | :------------------- | :--------------- | :----------
 | id                   | ID               | Identifier of the language from table below.
 | name                 | string           | Name of the language (might not match table below, e.g. if localized).
-| entry_point_required | boolean ?        | Whether the language requires an entry point. Defaults to `false`.
+| entry_point_required | boolean          | Whether the language requires an entry point. Defaults to `false`.
 | entry_point_name     | string ?         | The name of the type of entry point, such as "Main class" or "Main file"). Required iff entry_point_required is `true`.
 | extensions           | array of string  | File extensions for the language.
-| compiler             | Command object ? | Command used for compiling submissions.
-| runner               | Command object ? | Command used for running submissions. Relevant e.g. for interpreted languages and languages running on a VM.
+| compiler             | Command object   | Command used for compiling submissions.
+| runner               | Command object   | Command used for running submissions. Relevant e.g. for interpreted languages and languages running on a VM.
```

Comment:
I think for both of these, if the server declared intent to provide these, then it should provide it for *each* language, thus these properties should not be nullable.

Reply:
I get your point, this means you need to somehow know your version information before you can expose languages, which might mean you need to connect to some judgehost hardware to figure that out.

But you could of course just not send out the languages until then. Why should you not be able to import CMS data?

Reply:
> I get your point, this means you need to somehow know your version information before you can expose languages

No, you only need to know the `command`, all other fields are optional. And you *could* always put in a dummy value. It's exactly equally useless.

Reply:
Ah sorry, didn't see those fields were optional. Then yes, I think you'd know your command at the point you load the language, right? And if not put in something and change it later indeed.

Reply:
Sure, you can enter 'something' in any non-nullable property - but why require anyone do that, require provider to not send events until specific properties are added, or require manual interaction importing data from CMS? To me these are all additional requirements to solve a problem that doesn't exist. Nobody ever said 'don't tell me anything about a problem until it's 100% configured' or 'you should be forced to make up data and fix it later'. CCS can still decide to do that, but IMHO shouldn't be spec-required.

Reply:
The way I see it: if you already have languages, but somehow don't have the compiler/runner command data yet, then you can simply not add them to `/access` and not provide these properties. Only once you have them properly configured/loaded, you add them to `/access` and provide them in the languages endpoint.